src/istore_agg.c

```diff
@@ -59,11 +62,245 @@ state_extend(ISAggState *state, int n)
+
+/*
+ * Sum pairs into first one.
+ * Also does overflow check.  If the inputs are of different signs then their sum
+ * cannot overflow.  If the inputs are of the same sign, their sum had
+ * better be that sign too.
+ */
+#define SUM_PAIRS(pairs1, pairs2)                                       \
+    do {                                                                \
+        if (SAMESIGN(pairs1->val, pairs2->val))                         \
+        {                                                               \
+            pairs1->val += pairs2->val;                                 \
+            if (!SAMESIGN(pairs1->val, pairs2->val))                    \
+                ereport(ERROR,                                          \
+                        (errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),   \
+                         errmsg("bigint out of range")));               \
+        }                                                               \
+        else                                                            \
+            pairs1->val += pairs2->val;                                 \
+    } while (0)
+
+/* Internal SUM_FLOOR aggregation function for istore */
+static ISAggState *
+istore_agg_sum_floor_internal(ISAggState *state, IStore *istore)
+{
+    BigIStorePair *pairs1;
+    IStorePair    *pairs2;
+    int            index1 = 0, index2 = 0;
+    int            left;
+
+    pairs1 = state->pairs;
+    pairs2 = FIRST_PAIR(istore, IStorePair);
+    while (index1 < state->used && index2 < istore->len)
+    {
+        if (pairs1->key < pairs2->key)
+        {
+            // do nothing keep state
+            ++pairs1;
+            ++index1;
+        }
+        else if (pairs1->key > pairs2->key)
+        {
+            int i = 0,
+                c = 0;
+
+            // calculate how many new elements we have
+            while (index2 + i < istore->len && pairs1->key > pairs2[i].key)
+            {
+                if (pairs2[i].val >= state->ifloor)
+                    c++;
+
+                ++i;
+            }
+
+            if (c > 0)
+            {
+                // expand state capacity to store new items
+                state  = state_extend(state, c);
+                pairs1 = state->pairs + index1;
+
+                // move data c steps forward from index1
+                memmove(pairs1 + c, pairs1,
+                        (state->used - index1) * sizeof(BigIStorePair));
+                state->used += c;
+
+                for (int j = 0; j < i; j++)
+                {
+                    if (pairs2->val >= state->ifloor)
+                    {
+                        pairs1->key = pairs2->key;
+                        pairs1->val = pairs2->val;
+                        ++pairs1;
+                    }
+                    ++pairs2;
+                }
+                index1 += c;
+            }
+            else
+            {
+                // just move the position
+                pairs2 += i;
```

Comment:
It looks like what is it doing is skipping all keys with values below the floor one. This should only work for a particular case of the floor value being 0 and sum aggregate. Otherwise, with a non-zero floor value, we should change the actual value to the floor value and do the copy (in case the key is new) or sum (in case there is a match). Given that, I really think we should join it with the istore_agg_sum_internal, as the code is only marginally different and the istore_agg_sum_internal is already written to be a universal transition function for multiple aggregates.

Reply:
I don't understand why in case of non-zero floor we should change the value, the idea is skip all these values. And when we reach the value that is >= floor, we add a key first time.

Reply:
I think we should change and not skip them even in the case of a zero floor value, it's just that zero is ignored by the sum; my understanding of the floor function is that when applied to the elements of the istore it should returns another istore with all the keys lower than the floor boundary changed to be equal to that boundary, in other words `floor({key=>val}, N) = {key => val >= N ? val : N}`.

Reply:
Think, for instance, of an example of the revenue for 7 days a week, `{1=>100, 2=>-200, 3=>100, 4=>100, 5=>100, 6=>100, 7=>400}`; if we want to avoid having negatives in it, we would expect to change the value of the 2nd day to 0, but leave the key there, in other words:
`{1=>100, 2=>0, ...}`, so we would write istore_floor(istore, 0); now, given that, I don't think sum_floor should behave differently in respect to processing individual istores.